torch/utils/data/distributed.py

```diff
@@ -76,49 +73,37 @@ def __init__(self, dataset: Dataset, num_replicas: Optional[int] = None,
         self.rank = rank
         self.epoch = 0
         self.drop_last = drop_last
-        # If the dataset length is evenly divisible by # of replicas, then there
-        # is no need to drop any data, since the dataset will be split equally.
-        if self.drop_last and len(self.dataset) % self.num_replicas != 0:  # type: ignore
-            # Split to nearest available length that is evenly divisible.
-            # This is to ensure each rank receives the same amount of data when
-            # using this Sampler.
-            self.num_samples = math.ceil(
-                # `type:ignore` is required because Dataset cannot provide a default __len__
-                # see NOTE in pytorch/torch/utils/data/sampler.py
-                (len(self.dataset) - self.num_replicas) / self.num_replicas  # type: ignore
-            )
+        if self.drop_last:
+            self.num_samples = len(self.dataset) // self.num_replicas  # type: ignore
         else:
             self.num_samples = math.ceil(len(self.dataset) / self.num_replicas)  # type: ignore
         self.total_size = self.num_samples * self.num_replicas
         self.shuffle = shuffle
         self.seed = seed
 
-    def __iter__(self) -> Iterator[T_co]:
+    def __iter__(self) -> Iterator[int]:
         if self.shuffle:
             # deterministically shuffle based on epoch and seed
             g = torch.Generator()
             g.manual_seed(self.seed + self.epoch)
```

Comment:
Rather than having a separate method, `set_epoch`, to increment the random seed on each epoch, it would be easier (for the user) just to increment `self.seed` here in the iterator. Presumably the sampler isn't being iterated over a variable number of times, so the result should be the same without having the overhead of having to remember to increment the epoch. It's hard to see a situation where you _wouldn't_ want the samples to be reshuffled on each pass.